cs/src/Connections/TunnelRelayTunnelHost.cs

```diff
@@ -345,6 +351,61 @@ void RemoveClientSessionTask(Task t)
         }
     }
 
+    /// <summary>
+    /// Encrypts the channel if necessary when a port connection is established (v2 only).
+    /// </summary>
+    protected override void OnForwardedPortConnecting(
+        object? sender, ForwardedPortConnectingEventArgs e)
+    {
+        var channel = e.Stream.Channel;
+        var relayRequestMessage = channel.OpenMessage
+            .ConvertTo<PortRelayConnectRequestMessage>();
+
+        bool isE2EEncryptionEnabled = this.EnableE2EEncryption &&
+            relayRequestMessage.IsE2EEncryptionRequested;
+        if (isE2EEncryptionEnabled)
+        {
+            // Increase the max window size so that it is at least larger than the window
+            // size of one client channel.
+            channel.MaxWindowSize = SshChannel.DefaultMaxWindowSize * 2;
+
+            // TODO: Publish the host public key to the relay so that the client can verify.
+
+            SshServerCredentials serverCredentials =
+                new SshServerCredentials(HostPrivateKey);
+            var secureStream = new SecureStream(
+                e.Stream,
+                serverCredentials,
+                channel.Trace.WithName(channel.Trace.Name + "." + channel.ChannelId));
+
+            // The client was already authenticated by the relay.
+            secureStream.Authenticating += (_, e) =>
+                e.AuthenticationTask = Task.FromResult<ClaimsPrincipal?>(
+                    new ClaimsPrincipal());
+
+            e.TransformTask = Task.FromResult<Stream?>(secureStream);
+
+            // The client will connect to the secure stream after the channel is opened.
+            ConnectEncryptedChannel();
+            async void ConnectEncryptedChannel()
+            {
+                try
+                {
+                    // Do not pass the cancellation token from the connecting event,
+                    // because the connection will outlive the event.
+                    await secureStream.ConnectAsync();
+                }
+                catch (Exception ex)
+                {
+                    channel.Trace.Error(
```

Comment:
Let's catch any exceptions out of `channel.Trace.Error` to be safe in `async void` method.

Reply:
Done.